### ISSUE TITLE
Batch norm doc mentions using biased variance estimator, but it actually uses the unbiased variance estimator

### ISSUE THREAD

**Issue by zaccharieramzi** [pytorch/pytorch#77427]
### 📚 The doc issue

The [BatchNorm documentation](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html) clearly states:
> The standard-deviation is calculated via the biased estimator, equivalent to `torch.var(input, unbiased=False)`.

However, when looking at the computations, it appears to use the *unbiased* estimator as shown with the following code:

```python
import numpy as np
import torch
from torch.nn import BatchNorm2d, init

np.random.seed(0)
inputs_2d = np.random.normal(size=(16, 3, 8, 8)).astype(np.float32)

bn_momentum = 0.

offline_bn_torch = BatchNorm2d(3, eps=1e-18, momentum=1-bn_momentum)
init.constant_(offline_bn_torch.weight, 1)
init.constant_(offline_bn_torch.bias, 0)
offline_bn_torch.reset_running_stats()
print(list(offline_bn_torch.buffers())[1])
offline_bn_torch.train()
offline_bn_torch(torch.tensor(inputs_2d))
print(list(offline_bn_torch.buffers())[1])
inputs_2d = np.transpose(inputs_2d, (0, 2, 3, 1))
inputs_2d = np.reshape(inputs_2d, (-1, 3))
print(np.std(inputs_2d, axis=0)**2)
print(torch.var(torch.tensor(inputs_2d), 0, unbiased=True, keepdim=False))
print(torch.var(torch.tensor(inputs_2d), 0, unbiased=False, keepdim=False))
```

which outputs:
```
tensor([1., 1., 1.])
tensor([0.9247, 0.9724, 0.9275])
[0.9238283  0.9714646  0.92655975]
tensor([0.9247, 0.9724, 0.9275])
tensor([0.9238, 0.9715, 0.9266])
```

### Suggest a potential alternative/fix

The documentation should actually say:

> The standard-deviation is calculated via the unbiased estimator, equivalent to torch.var(input, unbiased=True).

I could submit a PR if needed.

cc @svekars @holly1238 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

**Comment by zaccharieramzi** [comment]
It turns out that after further verification, the problem is actually a bit more complex.
It was first noted in [this issue](https://github.com/pytorch/pytorch/issues/1410).

The estimator used to compute the variance is different in the 2 following situations:
- when doing the forward pass in train mode, the biased estimator (i.e. `std**2`) is used, it would correspond to `torch.var(input, unbiased=False)`.
- when storing the variance in the running variance buffer, to be later used in eval mode, the unbiased estimator is used (i.e. with the Bessel correction, multiplying by `n/(n-1)`), it would correspond to `torch.var(input, unbiased=True)`.

You can see that with the following code:
```
import numpy as np
import torch
from torch.nn import BatchNorm2d

np.random.seed(0)
batch_size = 16
im_size = 4
inputs = np.random.normal(size=(batch_size, 1, im_size, im_size)).astype(np.float32)
bn_torch = BatchNorm2d(1, eps=1e-7, momentum=0.1)
bn_torch.train()
out_bn_torch_train = bn_torch(torch.tensor(inputs)).detach().numpy()
mean_torch, var_torch, _ = [b.detach().cpu().numpy() for b in list(bn_torch.buffers())]
mean_np = np.mean(inputs)
var_np = np.std(inputs)**2
out_np = (inputs - mean_np) / np.sqrt(var_np + 1e-7)
np.testing.assert_allclose(out_bn_torch_train, out_np, rtol=1e-4, atol=1e-6)
out_torch = (inputs - mean_torch) / np.sqrt(var_torch + 1e-7)
np.testing.assert_allclose(out_bn_torch_train, out_torch, rtol=1e-4, atol=1e-6)  # will fail
bn_torch.eval()
out_bn_torch_eval = bn_torch(torch.tensor(inputs)).detach().numpy()
out_torch = (inputs - mean_torch) / np.sqrt(var_torch + 1e-7)
np.testing.assert_allclose(out_bn_torch_eval, out_torch, rtol=1e-4, atol=1e-6)
out_np = (inputs - mean_np) / np.sqrt(var_np + 1e-7)
np.testing.assert_allclose(out_bn_torch_eval, out_np, rtol=1e-4, atol=1e-6)  # will fail
```

I think this behaviour is very unexpected and should be clearly documented.
For example:
> At train time in the forward pass, the standard-deviation is calculated via the biased estimator, equivalent to `torch.var(input, unbiased=False)`.
However, the value stored in the moving average of the standard-deviation is calculated via the unbiased estimator, equivalent to `torch.var(input, unbiased=True)`.


Maybe it would be interesting also to look at the effect on training with small batches since [it was noted that this can have a significant impact](https://github.com/pytorch/pytorch/issues/1410#issuecomment-509989143).

**Comment by jbschlosser** [comment]
Hey @zaccharieramzi, thanks for the detailed report. I agree we should clarify what is going on in the docs wrt when Bessel's correction is applied. If you'd like to submit a PR doing so, I'm happy to review it.

**Comment by zaccharieramzi** [comment]
Yes I would be happy to submit a PR after my vacation (on the 20th).

**Comment by QIanGua** [comment]
> 



> It turns out that after further verification, the problem is actually a bit more complex. It was first noted in [this issue](https://github.com/pytorch/pytorch/issues/1410).
> 
> The estimator used to compute the variance is different in the 2 following situations:
> 
> * when doing the forward pass in train mode, the biased estimator (i.e. `std**2`) is used, it would correspond to `torch.var(input, unbiased=False)`.
> * when storing the variance in the running variance buffer, to be later used in eval mode, the unbiased estimator is used (i.e. with the Bessel correction, multiplying by `n/(n-1)`), it would correspond to `torch.var(input, unbiased=True)`.
> 
> You can see that with the following code:
> 
> ```
> import numpy as np
> import torch
> from torch.nn import BatchNorm2d
> 
> np.random.seed(0)
> batch_size = 16
> im_size = 4
> inputs = np.random.normal(size=(batch_size, 1, im_size, im_size)).astype(np.float32)
> bn_torch = BatchNorm2d(1, eps=1e-7, momentum=0.1)
> bn_torch.train()
> out_bn_torch_train = bn_torch(torch.tensor(inputs)).detach().numpy()
> mean_torch, var_torch, _ = [b.detach().cpu().numpy() for b in list(bn_torch.buffers())]
> mean_np = np.mean(inputs)
> var_np = np.std(inputs)**2
> out_np = (inputs - mean_np) / np.sqrt(var_np + 1e-7)
> np.testing.assert_allclose(out_bn_torch_train, out_np, rtol=1e-4, atol=1e-6)
> out_torch = (inputs - mean_torch) / np.sqrt(var_torch + 1e-7)
> np.testing.assert_allclose(out_bn_torch_train, out_torch, rtol=1e-4, atol=1e-6)  # will fail
> bn_torch.eval()
> out_bn_torch_eval = bn_torch(torch.tensor(inputs)).detach().numpy()
> out_torch = (inputs - mean_torch) / np.sqrt(var_torch + 1e-7)
> np.testing.assert_allclose(out_bn_torch_eval, out_torch, rtol=1e-4, atol=1e-6)
> out_np = (inputs - mean_np) / np.sqrt(var_np + 1e-7)
> np.testing.assert_allclose(out_bn_torch_eval, out_np, rtol=1e-4, atol=1e-6)  # will fail
> ```
> 
> I think this behaviour is very unexpected and should be clearly documented. For example:
> 
> > At train time in the forward pass, the standard-deviation is calculated via the biased estimator, equivalent to `torch.var(input, unbiased=False)`.
> > However, the value stored in the moving average of the standard-deviation is calculated via the unbiased estimator, equivalent to `torch.var(input, unbiased=True)`.
> 
> Maybe it would be interesting also to look at the effect on training with small batches since [it was noted that this can have a significant impact](https://github.com/pytorch/pytorch/issues/1410#issuecomment-509989143).

Hi @zaccharieramzi . I find that your code in eval mode part doesn't support the argument that `in eval mode, the unbiased estimator is used`, because  you only try prove  `unbiased!=Flase`. But `unbiased!=Flase` doesn't mean `unbiased=True` .  In fact,  at eval mode,  there is no sampling process , it just use the previous moving variance in train mode period, nothing more.  What's your opinion?
```
import torch
from torch.nn import BatchNorm2d
torch.manual_seed(0)
batch_size = 16
im_size = 4
inputs = torch.rand(batch_size, 1, im_size, im_size, dtype=torch.float32)
bn_torch = BatchNorm2d(1, eps=1e-7, momentum=0.1)

# train mode
bn_torch.train()
mean_torch = inputs.mean(dim=(0,2,3),keepdim=True)
var_torch = inputs.var(dim=(0,2,3),keepdim=True, unbiased=False)
bn0 = bn_torch(inputs) # in train mode, mean_torch and var_torch updates after applying bn_torch(inputs) each time
out_torch = (inputs - mean_torch) / (var_torch + 1e-7).sqrt()
print( torch.allclose(bn0, out_torch, rtol=1e-04, atol=1e-06) ) # true => biased variance
# eval mode
bn_torch.eval()
mean_torch, var_torch, _ = [b for b in list(bn_torch.buffers())] # keep unchanged in eval mode
out_torch = (inputs - mean_torch) / (var_torch + 1e-7).sqrt()
bn1 = bn_torch(inputs)
print( torch.allclose(bn1, out_torch, rtol=1e-04, atol=1e-06) ) # true
```

**Comment by monurcank** [comment]
`import torch

feature_size = 10
batch_size = 2000
batch_norm = torch.nn.BatchNorm1d(feature_size, momentum=1.0)

g_cpu = torch.Generator()
g_cpu.manual_seed(2147483647)
sample_input = torch.randn(batch_size, feature_size, generator=g_cpu)

batch_norm.train()
sample_output_train = batch_norm(sample_input)

batch_norm.eval()
sample_output_eval = batch_norm(sample_input)
print(torch.allclose(sample_output_train, sample_output_eval)) # False -> Train-test mismatch?

print(torch.var(sample_output_train, unbiased=True)) # 1.0000
print(torch.var(sample_output_eval, unbiased=True))  # 0.9995
# In train mode, the outputs are as expected since it uses biased variance.
# So, using biased variance is appropriate for our needs. We want the output to have a 1.0 variance.

# Therefore, eval mode shouldn't have a Bessel correction to have consistent results:
batch_norm_eval_mode_current = (sample_input - batch_norm.running_mean) / torch.sqrt(batch_norm.running_var + 1e-5)
batch_norm_eval_mode_should_be_like = (sample_input - batch_norm.running_mean) / torch.sqrt(batch_norm.running_var * (batch_size-1)/batch_size + 1e-5)
print(torch.allclose(sample_output_eval, batch_norm_eval_mode_current)) # True
print(torch.allclose(sample_output_train, batch_norm_eval_mode_should_be_like)) # True
# Both estimators must have the same statistical properties.

print("----------------------------------------------------------------")
my_calculation_unbiased = (sample_input - torch.mean(sample_input, dim=0)) / torch.sqrt(torch.var(sample_input, dim=0, unbiased=False) + 1e-5)
print(torch.allclose(sample_output_train, my_calculation_unbiased)) # True
print(torch.allclose(sample_output_eval, my_calculation_unbiased)) # False

my_calculation_biased = (sample_input - torch.mean(sample_input, dim=0)) / torch.sqrt(torch.var(sample_input, dim=0, unbiased=True) + 1e-5)
print(torch.allclose(sample_output_train, my_calculation_biased)) # False
print(torch.allclose(sample_output_eval, my_calculation_biased)) # True
`

**Comment by zaccharieramzi** [comment]
@QIanGua I am not sure what you mean by "at eval mode, there is no sampling process ". The Bessel correction does not depend on whether there is a sampling process, it's just a matter of using correction factor or not (`n/(n-1)`).

For the eval mode I am doing the 2 following tests:
`np.testing.assert_allclose(out_bn_torch_eval, out_torch, rtol=1e-4, atol=1e-6)` (this tests whether we use the unbiased estimator obtained from the parameters of the batch norm, and it passes) and `np.testing.assert_allclose(out_bn_torch_eval, out_np, rtol=1e-4, atol=1e-6)` (this tests whether we use the biased estimator obtained from the numpy computation, and it fails).

Basically I don't understand the following: 
> But `unbiased!=False` doesn't mean `unbiased=True `
